perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -408,6 +408,178 @@ public Group updateGroup(PerunSession sess, Group group) throws InternalErrorExc
+					getGroupsManagerImpl().updateGroupName(sess, gr);
+
+					// create auditer message for every updated group
+					getPerunBl().getAuditer().log(sess, "{} updated.", gr);
+
+					levelSubGroups.addAll(getSubGroups(sess, gr));
+				}
+				subGroups.clear();
+				subGroups.addAll(levelSubGroups);
+			}
+
+			// And finally update parentGroupId for moving group in database
+			this.updateParentGroupId(sess, movingGroup);
+
+			//After that, we have to move indirect members as well
+			List<Member> members = getGroupMembers(sess, movingGroup);
+
+			try {
+				//Removing indirect members from old position
+				processRelationMembers(sess, groupsManagerImpl.getGroupById(sess, oldParentGroupId), members, movingGroup.getId(), false);
+				//Adding indirect members to new position
+				processRelationMembers(sess, groupsManagerImpl.getGroupById(sess, movingGroup.getParentGroupId()), members, movingGroup.getId(), true);
+			} catch (GroupOperationsException e) {
+				throw new InternalErrorException("Some operation was not allowed during processing members relations", e);
+			} catch (GroupNotExistsException e) {
+				throw new InternalErrorException("Some group does not exists...", e);
+			}
+		} else {
+
+			// check if moving group is already under destination group
+			if (movingGroup.getParentGroupId() == null) {
+				throw new GroupMoveNotAllowedException("Moving group: " + movingGroup + " is already top level group.");
+			}
+
+			List<Group> destinationGroupSubGroups = getGroups(sess, getVo(sess, movingGroup));
+
+			// check if under destination group is group with same short name as Moving group short name
+			for (Group group: destinationGroupSubGroups) {
+				if(movingGroup.getShortName().equals(group.getName())){
+					throw new GroupMoveNotAllowedException("There is already top level group with the same name as moving group: " + movingGroup + ".");
+				}
+			}
+
+			//Save old parent id for removing indirect members
+			int oldParentGroupId = movingGroup.getParentGroupId();
+			// We can move whole moving group tree as top level group in vo
+			movingGroup.setParentGroupId(null);
+			movingGroup.setName(movingGroup.getShortName());
+			//we have to update group name in database
+			getGroupsManagerImpl().updateGroupName(sess, movingGroup);
+
+			List<Group> subGroups = getSubGroups(sess, movingGroup);
+			List<Group> levelSubGroups = new ArrayList<>();
+
+			// We have to properly set all subGroups names level by level
+			while (!subGroups.isEmpty()) {
+				levelSubGroups.clear();
+				for (Group gr : subGroups) {
+					try {
+						gr.setName(getParentGroup(sess, gr).getName() + ":" + gr.getShortName());
+					} catch (ParentGroupNotExistsException ex) {
+						//that should never happen
+						throw new InternalErrorException("Parent group does not exists for group " + gr);
+					}
+					//we have to update each subGroup name in database
+					getGroupsManagerImpl().updateGroupName(sess, gr);
```

Comment:
You are missing audit message about group update ``getPerunBl().getAuditer().log(sess, "{} updated.", gr);`` after this line, since all subgroups were changed, like in a case of moving group to top-level.